LLVM 20

alexrp opened the pull request:
Currently targeting LLVM 20.1.2.

Test suite status:

* `test-fmt`: :heavy_check_mark:
* `test-cases`: :heavy_check_mark:
* `test-translate-c`: :heavy_check_mark:
* `test-run-translated-c`: :heavy_check_mark:
* `test-behavior`: :heavy_check_mark:
* `test-c-import`: :heavy_check_mark:
* `test-compiler-rt`: :heavy_check_mark:
* `test-universal-libc`: :heavy_check_mark:
* `test-std`: :heavy_check_mark:
* `test-unit`: :heavy_check_mark:
* `test-compare-output`: :heavy_check_mark:
* `test-standalone`: :heavy_check_mark:
* `test-c-abi`: :heavy_check_mark:
* `test-link`: :heavy_check_mark:
* `test-stack-traces`: :heavy_check_mark:
* `test-cli`: :heavy_check_mark:
* `test-asm-link`: :heavy_check_mark:
* `test-debugger`: :heavy_check_mark:
* `test-incremental`: :heavy_check_mark:

Bugs filed upstream:

* https://github.com/llvm/llvm-project/issues/125954
* https://github.com/llvm/llvm-project/issues/127296
* https://github.com/llvm/llvm-project/issues/127298
* https://github.com/llvm/llvm-project/issues/129843
* https://github.com/llvm/llvm-project/issues/130408
* https://github.com/llvm/llvm-project/issues/130541

Closes #22014.

Previously: https://github.com/ziglang/zig/pull/21183

alexrp committed:
ci: Stop triggering the workflow for llvm branches.

We usually do a PR for the llvm branches, so triggering a run for the branch
means we do extra work for no reason.

alexrp committed:
build: Update to LLVM/Clang 20.

alexrp committed:
zig cc: Update driver files to Clang 20.

alexrp committed:
zig cc: Update intrinsic headers to Clang 20.

alexrp committed:
libunwind: Update to LLVM 20.

alexrp committed:
libcxxabi: Update to Clang 20.

alexrp committed:
libcxx: Update to Clang 20.

See:

* https://discourse.llvm.org/t/rfc-freezing-c-03-headers-in-libc/77319
* https://discourse.llvm.org/t/rfc-project-hand-in-hand-llvm-libc-libc-code-sharing/77701

We're dropping support for C++03 for Zig due to the first change; it would be
insane to ship 1018 duplicate header files just for this outdated use case.

As a result of the second change, I had to bring in a subset of the headers from
llvm-libc since libc++ now depends on these. Hopefully we can continue to get
away with not copying the entirety of llvm-libc.

alexrp committed:
compiler: Updates for LLVM/Clang 20 API changes.

alexrp committed:
std.Target: Update CPU models/features for LLVM 20.

Closes #21818.

alexrp committed:
zig cc: Update options data to Clang 20.

alexrp committed:
libtsan: Update to LLVM 20.

alexrp committed:
llvm: Set target-abi module flag.

LLVM is increasingly making use of this module flag when present.

alexrp committed:
llvm: Fix i128 alignment for various targets.

This ABI bug was fixed in LLVM 20.

alexrp committed:
compiler: Update max int alignments for some targets.

alexrp committed:
compiler-rt: Fix names of some float/int conversion routines for hexagon.

alexrp commented:
Linked gists have been updated. Current status:

`test-behavior` only has a couple of failures with the self-hosted backend now. It looks like either an LLVM miscompilation in compiler-rt, or undefined behavior in compiler-rt or the relevant parts of the standard library. Here's a better stack trace:

```
❯ ./zig-out/bin/zig test -fno-llvm test/behavior/x86_64/math.zig --test-filter intFromFloat --debug-rt
Illegal instruction at address 0x1486214
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/debug/no_panic.zig:25:5: 0x1486214 in outOfBounds (compiler_rt)
    @trap();
    ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/mem.zig:1986:20: 0x148ae58 in writePackedIntLittle__anon_4353 (compiler_rt)
        write_bytes[store_size] &= ~((@as(u8, 1) << @as(u3, @intCast(tail_len))) - 1);
                   ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/compiler_rt/int_from_float.zig:100:33: 0x1489e3f in __fixdfei (compiler_rt)
    std.mem.writePackedIntNative(I, std.mem.sliceAsBytes(result), exponent, int);
                                ^
/home/alexrp/Source/ziglang/zig-llvm20/test/behavior/x86_64/math.zig:22099:12: 0x111e0de in testArgKinds__anon_9357 (math.zig)
    return @intFromFloat(rhs);
           ^
Unwind error at address `exe:0x111e0de` (error.AddressOutOfRange), trace may be incomplete

/home/alexrp/Source/ziglang/zig-llvm20/test/behavior/x86_64/math.zig:1774:29: 0x111e562 in testArgs__anon_9353 (math.zig)
            try testArgKinds(
                            ^
/home/alexrp/Source/ziglang/zig-llvm20/test/behavior/x86_64/math.zig:9245:25: 0x11e9269 in testIntsFromFloats (math.zig)
            try testArgs(i256, f64, -0x0.8p256);
                        ^
/home/alexrp/Source/ziglang/zig-llvm20/test/behavior/x86_64/math.zig:22103:47: 0x12eeca7 in decltest.intFromFloat (math.zig)
    try test_int_from_float.testIntsFromFloats();
                                              ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/compiler/test_runner.zig:214:25: 0x1433e1b in mainTerminal (test_runner.zig)
        if (test_fn.func()) |_| {
                        ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/compiler/test_runner.zig:62:28: 0x142a6d5 in main (test_runner.zig)
        return mainTerminal();
                           ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/start.zig:647:22: 0x1423b9e in posixCallMainAndExit (std.zig)
            root.main();
                     ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/start.zig:271:5: 0x1423481 in _start (std.zig)
    asm volatile (switch (native_arch) {
    ^
error: the following test command crashed:
/home/alexrp/Source/ziglang/zig-llvm20/.zig-cache/o/18c5968e5d1bf3f135c3c144c37b031f/test --seed=0x1f613aa8
```

The `test-std` failures for `aarch64_be` and `mips64(el)` look weird. The former might be LLVM miscompiling `@cmpxchgStrong`? No clue about the latter yet.

The `test-link` failure seems like it could just be a bug on our side. I'll be investigating this one next.

No idea about the `test-stack-traces` failure yet, but these tests tend to be a bit flaky between LLVM versions. It shouldn't be too hard to investigate.

I think once LLVM 20.1.1 is out, we can put it on the CI machines and start testing this PR.

alexrp commented:
@mlugg IIRC you worked on error trace stuff recently. Do the `test-stack-traces` failures ring a bell?

mlugg commented:
> @mlugg IIRC you worked on error trace stuff recently. Do the `test-stack-traces` failures ring a bell?

They don't ring a bell, but: the only way I can see those failures happening is `@returnAddress()` returning a bad result (perhaps address 0). Specifically, it's the call to `@returnAddress` in `std.builtin.returnError`. That's the function responsible for pushing frames onto the error return trace, and I'm pretty confident the logic for calling it and that function itself are all fine. So, maybe take a look at the behavior there in a debugger and see what the address is?

alexrp committed:
Sema: Prevent tail calls of std.builtin.returnError().

LLVM 20 started tail-calling it in some of our test cases, resulting in:

    error: AndMyCarIsOutOfGas
    /home/alexrp/Source/ziglang/zig-llvm20/repro.zig:2:5: 0x103ef9d in main (repro)
        return error.TheSkyIsFalling;
        ^
    /home/alexrp/Source/ziglang/zig-llvm20/repro.zig:6:5: 0x103efa5 in main (repro)
        return error.AndMyCarIsOutOfGas;
        ^
    /home/alexrp/Source/ziglang/zig-llvm20/lib/std/start.zig:656:37: 0x103ee83 in posixCallMainAndExit (repro)
                const result = root.main() catch |err| {
                                        ^

instead of the expected:

    error: AndMyCarIsOutOfGas
    /home/alexrp/Source/ziglang/zig-llvm20/repro.zig:2:5: 0x103f00d in main (repro)
        return error.TheSkyIsFalling;
        ^
    /home/alexrp/Source/ziglang/zig-llvm20/repro.zig:6:5: 0x103f015 in main (repro)
        return error.AndMyCarIsOutOfGas;
        ^
    /home/alexrp/Source/ziglang/zig-llvm20/repro.zig:11:9: 0x103f01d in main (repro)
            try bar();
            ^

alexrp committed:
llvm: never_tail implies never_inline, so set noinline in this case too.

alexrp committed:
std: Disable `link with relative paths` and `linkat with different directories` tests for mips64(el).

These started failing with LLVM 20 for unclear reasons:

test-std
└─ run test std-mips64-linux.4.19...6.13.4-gnuabi64.2.28-mips64r2-Debug-libc 2798/2878 passed, 2 failed, 78 skipped
error: 'posix.test.test.link with relative paths' failed: expected 2, found 0
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/testing.zig:103:17: 0x1d9e5bf in expectEqualInner__anon_47031 (test)
                return error.TestExpectedEqual;
                ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/posix/test.zig:311:9: 0x3650f57 in test.link with relative paths (test)
        try testing.expectEqual(@as(@TypeOf(nstat.nlink), 2), nstat.nlink);
        ^
error: 'posix.test.test.linkat with different directories' failed: expected 2, found 0
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/testing.zig:103:17: 0x1d9e5bf in expectEqualInner__anon_47031 (test)
                return error.TestExpectedEqual;
                ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/posix/test.zig:355:9: 0x3653377 in test.linkat with different directories (test)
        try testing.expectEqual(@as(@TypeOf(nstat.nlink), 2), nstat.nlink);
        ^
error: while executing test 'zig.system.darwin.macos.test.detect', the following test command failed:
qemu-mips64 -L /opt/glibc/mips64-linux-gnu-n64 /home/alexrp/Source/ziglang/zig-llvm20/.zig-cache/o/22a8c3762ea56ae3a674fa9ad15f6657/test --seed=0xa1dbb43c --cache-dir=/home/alexrp/Source/ziglang/zig-llvm20/.zig-cache --listen=-

test-std
└─ run test std-mips64-linux.4.19...6.13.4-gnuabi64.2.28-mips64r2-Debug-libc 2798/2878 passed, 1 failed, 79 skipped
error: 'posix.test.test.linkat with different directories' failed: expected 2, found 0
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/testing.zig:103:17: 0x1d9e22f in expectEqualInner__anon_47031 (test)
                return error.TestExpectedEqual;
                ^
/home/alexrp/Source/ziglang/zig-llvm20/lib/std/posix/test.zig:356:9: 0x3650b47 in test.linkat with different directories (test)
        try testing.expectEqual(@as(@TypeOf(nstat.nlink), 2), nstat.nlink);
        ^
error: while executing test 'zig.system.darwin.macos.test.detect', the following test command failed:
qemu-mips64 -L /opt/glibc/mips64-linux-gnu-n64 /home/alexrp/Source/ziglang/zig-llvm20/.zig-cache/o/22a8c3762ea56ae3a674fa9ad15f6657/test --seed=0xa1dbb43c --cache-dir=/home/alexrp/Source/ziglang/zig-llvm20/.zig-cache --listen=-

Unfortunately, neither GDB nor LLDB want to play nice with qemu-mips64(el) at
the moment, so I can't easily debug these failures.

alexrp committed:
test: mips32 is no longer a slow target with LLVM 20.

We can now run these tests as part of test-modules w/o -Dtest-slow-targets.

Closes #21096.

alexrp commented:
This is now green on `x86_64-linux` with `-fqemu -fwasmtime --glibc-runtimes`. Once LLVM 20.1.1 is out, we should hopefully be able to revert ba2c47e1d1e41f7e93335a8cc6b64e3afcd38884 and 7f53cb248b60b9b7893c1a620428da8bf9191b58, and then put that LLVM version on the CI machines to see what the situation on macOS and Windows looks like.

alexrp committed:
llvm: Update the list of targets that use native f16/f128.

Closes #22003.
Closes #22013.

alexrp committed:
std.Target: Bump baseline hexagon model to hexagonv68.

https://github.com/llvm/llvm-project/pull/125584

alexrp committed:
llvm: Allow FastISel on mips again.

Closes #21215.

alexrp committed:
Revert "compiler: Don't consider powerpc to have red zone support yet."

This reverts commit 4fad60fd3a70d0b059ce92ce825faabc1d2ac2e8.

Closes #23056.

alexrp committed:
llvm: Use muslabin32/muslabi64 environments in the target triple.

Closes #2909.

alexrp committed:
Revert "test: Partially disable `vector float operators` on mips."

This reverts commit 19a91084c2a29175bad47d41cd11935904ede0b9.

Closes #21051.

alexrp committed:
Revert "glibc: Patch to work around missing features in LLVM's s390x assembler."

This reverts commit b230e4f598bd18b47f3f1c981869c597a06c7452.

Closes #21329.

alexrp committed:
llvm: Remove workaround for zero-length memset/memcpy on wasm.

Closes #16360.

alexrp committed:
std.Target: Add Abi.muslf32 and Abi.muslsf.

alexrp committed:
Compilation: Remove the PowerPC soft float preprocessor workaround.

Closes #21411.

alexrp committed:
zig cc: Avoid passing any "CPU" features related to float ABI to Clang.

alexrp committed:
std.Target: Bump baseline bpfel/bpfeb model to v3.

https://github.com/llvm/llvm-project/pull/131691

alexrp committed:
build: Rename -Dtest-slow-targets to -Dtest-extra-targets.

This can be used more broadly for targets that aren't quite ready to be tested
by default yet.

andrewrk commented:
I think you accidentally destroyed the commit I just pushed:

```diff
commit 69126d08c907724fe1c6713a37a536297fc53674 (HEAD -> llvm20, origin/llvm20)
Author: Andrew Kelley <andrew@ziglang.org>
Date:   Wed Apr 2 17:44:26 2025 -0700

    build.zig: add missing llvm libs

diff --git a/build.zig b/build.zig
index b298f87f9c..830cae2702 100644
--- a/build.zig
+++ b/build.zig
@@ -1196,6 +1196,10 @@ const llvm_libs = [_][]const u8{
     "LLVMSystemZCodeGen",
     "LLVMSystemZDesc",
     "LLVMSystemZInfo",
+    "LLVMSPIRVCodeGen",
+    "LLVMSPIRVDesc",
+    "LLVMSPIRVInfo",
+    "LLVMSPIRVAnalysis",
     "LLVMSparcDisassembler",
     "LLVMSparcAsmParser",
     "LLVMSparcCodeGen",

```

alexrp commented:
I just folded it into 071211548ea4f36eb74d09bb4987877d0fe5dd9e, along with a corresponding change to `cmake/FindLlvm.cmake`.

andrewrk commented:
Ah perfect, thanks!

alexrp committed:
ci: Update tarballs to 0.15.0-dev.203+53270f008.

alexrp committed:
compiler: Recognize -fno-sanitize=<...> in addition to -fsanitize=<...>.

alexrp committed:
cmake: Pass -fno-sanitize=undefined when building zig2.

We build zig2.c and compiler_rt.c with -O0 but then proceed to link with -O3.
So zig2.o and compiler_rt.o will have references to ubsan-rt symbols, but the
-O3 causes the compiler to not link ubsan-rt. We don't actually need the safety
here, so just explicitly disable ubsan.

alexrp committed:
test: Disable `error union switch with call operand` (ReleaseSafe) on macos.

This started failing in LLVM 20:

test
+- test-stack-traces
   +- check error union switch with call operand (ReleaseSafe llvm) failure
error:
========= expected this stdout: =========
error: TheSkyIsFalling
source.zig:3:5: [address] in [function]
    return error.TheSkyIsFalling;
    ^

========= but found: ====================
error: TheSkyIsFalling
source.zig:13:27: [address] in [function]
        error.NonFatal => return,
                          ^

andrewrk committed:
ci: update tarballs to 0.15.0-dev.203+53270f008

these have assertions enabled

andrewrk reviewed (approved):
Fantastic work.

This makes me wonder if we should define `noinline` to also disable tail calls. :thinking:
